extension/spring-test/src/main/kotlin/org/camunda/bpm/extension/process_test_coverage/spring_test/ProcessEngineCoverageTestExecutionListener.kt

```diff
@@ -0,0 +1,221 @@
+package org.camunda.bpm.extension.process_test_coverage.spring_test
+
+import mu.KLogging
+import org.assertj.core.api.Assertions
+import org.assertj.core.api.Condition
+import org.camunda.bpm.engine.ProcessEngine
+import org.camunda.bpm.engine.impl.cfg.ProcessEngineConfigurationImpl
+import org.camunda.bpm.extension.process_test_coverage.engine.CompensationEventCoverageHandler
+import org.camunda.bpm.extension.process_test_coverage.engine.ElementCoverageParseListener
+import org.camunda.bpm.extension.process_test_coverage.engine.ExecutionContextModelProvider
+import org.camunda.bpm.extension.process_test_coverage.model.DefaultCollector
+import org.camunda.bpm.extension.process_test_coverage.model.Run
+import org.camunda.bpm.extension.process_test_coverage.model.Suite
+import org.camunda.bpm.extension.process_test_coverage.util.CoverageReportUtil
+import org.springframework.core.Ordered
+import org.springframework.test.context.TestContext
+import org.springframework.test.context.TestExecutionListener
+
+/**
+ * Test execution listener for process test coverage.
+ * Can be used with spring testing framework to get process test coverage in spring tests.
+ *
+ * @author Jan Rohwer
+ */
+class ProcessEngineCoverageTestExecutionListener : TestExecutionListener, Ordered {
+
+    companion object : KLogging() {
+        /**
+         * If you set this property to a ratio (e.g. "1.0" for full coverage),
+         * the Extension will fail the test run if the coverage is less.<br></br>
+         * Example parameter for running java:<br></br>
+         * `-Dorg.camunda.bpm.extension.process_test_coverage.ASSERT_AT_LEAST=1.0`
+         */
+        const val DEFAULT_ASSERT_AT_LEAST_PROPERTY = "org.camunda.bpm.extension.process_test_coverage.ASSERT_AT_LEAST"
+    }
+
+    /**
+     * The state of the current run (class and current method).
+     */
+    private val coverageCollector = DefaultCollector(ExecutionContextModelProvider())
+
+    /**
+     * Conditions to be asserted on the class coverage percentage.
+     */
+    private val classCoverageAssertionConditions: MutableList<Condition<Double>> = mutableListOf()
+
+    private lateinit var processEngineCoverageProperties: ProcessEngineCoverageProperties
+
+    private var suiteInitialized = false
+
+    private fun loadConfiguration(testContext: TestContext) {
+        processEngineCoverageProperties = testContext.applicationContext.getBean(ProcessEngineCoverageProperties::class.java)
+        processEngineCoverageProperties.coverageAtLeast()?.let { addClassCoverageAtLeast(it) }
+        System.getProperty(processEngineCoverageProperties.optionalAssertCoverageAtLeastProperty())?.let {
+            try {
+                addClassCoverageAtLeast(it.toDouble())
+            } catch (e: NumberFormatException) {
+                throw RuntimeException("BAD TEST CONFIGURATION: system property \"${processEngineCoverageProperties.optionalAssertCoverageAtLeastProperty()}\" must be double")
```

Comment:
I don't like `RuntimeException` as a thrown type. It is a base type for a whole class of exceptions. A more specific exception is desired (IllegalArgument) in this case...

And I don't like the way you are doing it here. Parsing a system property to double deserves a separate function, please.

Reply:
Moved logic into function in builder, which now throws an IllegalArgumentException